bigbluebutton-html5/imports/ui/components/modals/settings/SettingsModal.jsx

```diff
@@ -53,36 +67,129 @@ export default class SettingsModal extends BaseModal {
     return <Submenu {...props}/>;
   }
 
+  /* When an option in the menu is clicked, set the activeSubmenu and focusSubmenu
+   * to the value of index. If clicked out of bounds set to 0 or end of submenus array accordingly.
+   *
+   * activeSubmenu: the submenu to be displayed to the user
+   * focusSubmenu: the submenu to set focus to
+   */
   clickSubmenu(i) {
-    if (this.submenus[i].className != SessionMenu) {
-      this.setState({ activeSubmenu: i });
-    };
+    if (i <= 0) {
+      this.setState({ activeSubmenu: 0, focusSubmenu: 0, });
+      return;
+    }
+
+    if (i >= this.submenus.length) {
+      this.setState({ activeSubmenu: this.submenus.length - 1,
+        focusSubmenu: this.submenus.length - 1, });
+      return;
+    } else {
+      this.setState({ activeSubmenu: i, focusSubmenu: i, });
+    }
   }
 
-  doubleClickSubmenu(i) {
-    if (this.submenus[i].className == SessionMenu) {
-      this.setState({ activeSubmenu: i });
+  /* Checks for key presses within the submenu list. Key behaviour varies.
+   *
+   * Tab: changes focus to next submenu or element outside of menu
+   * Shift+Tab: changes focus to previous submenu or element outside of menu
+   * Up Arrow: changes focus to previous submenu, can cycle through menu
+   * Down Arrow: changes focus to next submenu, can cycle through menu
+   * Spacebar: selects submenu in focus and sets as active
+   * Enter: selects submenu in focus and sets as active
+   */
+  handleKeyDown(event) {
+    // tab
+    if (event.keyCode === 9) {
+      let newIndex = 0;
+      if (this.state.focusSubmenu >= this.submenus.length - 1) {
+        newIndex = this.submenus.length - 1;
+      } else {
+        newIndex = this.state.focusSubmenu + 1;
+      }
+
+      this.setState({ focusSubmenu: newIndex });
+      return;
     }
+
+    // shift+tab
+    if (event.shiftKey && event.keyCode === 9) {
+      let newIndex = 0;
+      if (this.state.focusSubmenu <= 0) {
+        newIndex = 0;
+      } else {
+        newIndex = this.state.focusSubmenu - 1;
+      }
+
+      this.setState({ focusSubmenu: newIndex });
+      return;
+    }
+
+    // up arrow
+    if (event.keyCode === 38) {
+      if (this.state.focusSubmenu <= 0) {
+        this.setState({ focusSubmenu: this.submenus.length - 1 }, function () {
+          ReactDOM.findDOMNode(this.refs[`submenu${this.state.focusSubmenu}`]).focus();
```

Comment:
Destructuring `findDOMNode` would make the code cleaner by avoiding `ReactDOM` repetition